Remove nested setlisteners and setlisteners out of timer loops

wlbragg opened the pull request:
Fixes #1265 

This appears to completely fix the inconsistent nasal errors and segfaults that started sometime after the FGFS tagging of version 2019.

For the most part it remove nested setlisteners and setlisteners out of timer loops as well as reworking the mooring re-init to depend on a preexisting flag.

High priority for merging! Please test and verify and then merge.

The only outstanding code is the immat.nas nested listeners that Richard @Zaretto suggested removing. I need verification as to what this code is for and what we need to do to it.

wlbragg commented:
I went ahead and removed the setlistener to outside the fdm-initialized setlistener. Everything appears to still work, the call sign works as it is supposed to, so I think this is good to go!

wlbragg commented:
After this is merged I will merge master into frost and fog and verify it is working correctly and it should be ready to merge after that.

wlbragg commented:
This is ready to merge as far as I am concerned. It cleans up the nasal errors and segfaults on re-init.
It temporarily removes the planing-lift-lbs property from the ``` hydro/active-norm ``` allowing the floats and amphibious to work again.

wlbragg commented:
@legoboyvdlp @gilbertohasnofb 
The tiedowns-can-have property limits certain gear undercarriages from being able to use the tie downs. Anyone know why the bushkit 3 (floats) is not allowed to have tie-downs? I understand no chocks for greater than 2, but the tie-downs should work on all aircraft variants, yes?

dany93 commented:
Tests with Issue-1265 branch:
PHNL
- 1 - "Spawn at seaport if available unchecked"
On the runway.
(Default gear), Checked "Amphibious"--> (160 hp to) 180 hp (correct)
"Go Now" --> no effect, and these nasal errors:
```
"Go Now"

  197.59 [ALRT]:nasal      PORT PHNL    Index 2
  197.59 [ALRT]:nasal       LAT 21.31699999999999 LON -157.912 HEAD 190
  197.59 [ALRT]:nasal      Nasal runtime error: non-objects have no members
  197.59 [ALRT]:nasal        at /home/dany/FG/github/c172p/Nasal/mooring.nas, line 117
  197.59 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 143
  197.59 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 127
  197.59 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 96
  197.59 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 89
  197.59 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 76
  197.59 [ALRT]:nasal        called from: /home/dany/fgfs-D/install/flightgear/fgdata/Nasal/globals.nas, line 119
```
Takeoff from the runway, "landing" on the water, takeoff again, everything correct, wake appears and disappears as expected.

- 2 -  With "Spawn at seaport if available checked" prior to this launching.
It spawns **on the runway**,
Amphibious with **160hp** is checked, (I can check 180hp to takeoff)
 with these nasal errors:

```
With "Spawn at seaport if available" pre-checked

   58.29 [ALRT]:nasal      PORT PHNL    Index 2
   58.29 [ALRT]:nasal       LAT 21.31699999999999 LON -157.912 HEAD 190
   58.29 [ALRT]:nasal      Nasal runtime error: non-objects have no members
   58.29 [ALRT]:nasal        at /home/dany/FG/github/c172p/Nasal/mooring.nas, line 117
   58.29 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 143
   58.29 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 127
   58.29 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 96
   58.29 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 89
   58.29 [ALRT]:nasal        called from: /home/dany/FG/github/c172p/Nasal/mooring.nas, line 40

```
Is it the expected temporary behavior?
On the water and at takeoff again, everything seems correct (it floats, it does planing, wake).
But "Go Now" or "Spawn at seaport if available" do nothing (except nasal errors).

dany93 commented:
I did a "**relocatio**n" from LFMN to PHNL.
No crash, no nasal errors.
PHNL:
"Go Now" is greyed, not available (no effect).
Parking brake is "ON", no way to set it "OFF" (neither by Shift-b, nor by the handle, nor by props set)
(Ctrl-U to takeoff)
"Landing" on water, takeoff, wake, everything correct.

wlbragg commented:
>"Go Now" is greyed, not available (no effect).

I think I fixed that in the j3cub, I'll review and see if this is the same thing.

>Tests with Issue-1265 branch:
>PHNL

>    1 - "Spawn at seaport if available" unchecked
    On the runway.
    (Default gear), Checked "Amphibious"--> (160 hp to) 180 hp (correct)
    "Go Now" --> no effect, and these nasal errors:

I don't think I even tested that. I'll take a look, thanks!

dany93 commented:
> PHNL
>     1 - "Spawn at seaport if available" unchecked
>     On the runway, with Amphibious.
>    (I have to change 160 hp to) 180 hp ==> **not normal**
>     "Go Now" --> no effect, and these nasal errors:
> 

I confirm, even after waiting about 1 mn on the PHNL runway.
```
/controls/mooring/allowed = 1
/controls/mooring/port-available = true
(others = false)

```
By clicking "Go Now"
`/controls/mooring/go-to-moorings` changes to 'true' , but no visible effect and nasal errors.

wlbragg commented:
@dany93 I think the nasal error was the issue with all your other issues. I know what that was and corrected it.

I also changed the "saved data" ie: bushkit. I removed it from the .set saved-data code and added it to the mooring relocation code.
What this change is all about...
Our save aircraft state is still influx and needs a major overhaul to use the "Startup state" system effectively. I know what we need to do but it is serious amount of work. In the meantime we are using a scaled down version of it, for lack of better words. Long story short, the default saved state, variant selection and saved aircraft data all needs to work together through a logical flow of options, I really need to map it out in a flowchart in order to explain what I think needs to be done. But for now, our variant selections work if we don't save to aircraft-data the value of bushkit. That is why it was removed fron the .set file saved-data block. But that caused trouble on the mooring relocation as it wouldn't retain the float/amph choice because bushkit was no longer cached. So I moved it back to the saved-data block .set file. That broke the variant selection. So I finally had the idea to only save the bushkit prop on mooring relocation. It seems to work as I want it to.
You may want to also check briefly the variant selections mixed in with the mooring relocation. I did a quick test and it did what I was expecting.

wlbragg commented:
FYI: I found the actual culprit causing all the segfaults. 

It was including
``` light_manager.init();```
inside of
``` setlistener("/sim/signals/fdm-initialized", func { ...} ```

None of the other changes were causing obvious trouble. I guess we should leave the changes in place as they are technically a better way to do them?

wkitty42 commented:
On 4/19/19 4:41 PM, wlbragg wrote:
> FYI: I found the actual culprit causing all the segfaults.

what was different about it compared to all the others?

are you saying that it kept firing the light_manager.init() function? what does 
that function do?

dany93 commented:
Tests at PHNL
Amphibious and Floats
- With "Go Now"
- With "Spawn at seaport if available"
(for each).

Everything goes well.
It spawns close to a mooring.
Taking off, particles appear correctly on the water and disappear in flight, landing on the water.

Several non deleterious messages like
```
In file /home/dany/FG/github/c172p/Systems/c172p-hydrodynamics.xml: line 47
      Property hydro/float-beam-ft2 is already defined.
In file /home/dany/FG/github/c172p/Systems/c172p-heat.xml: line 22
      Property heat/human-body-temp-degc is already defined.
In file /home/dany/FG/github/c172p/Systems/indicated-airspeed.xml: line 45
      Property /instrumentation/airspeed-indicator/indicated-speed-kt is already defined.
In file /home/dany/FG/github/c172p/Systems/mooring-jsbsim.xml: line 81
      Property mooring/mooring-connected is already defined.
```
Still this 160 hp issue with Floats or Amphibious at launching. Is it difficult to fix? Is it connected to

> Our save aircraft state is still influx and needs a major overhaul to use the "Startup state" system effectively.
> (...)
>  But for now, our variant selections work if we don't save to aircraft-data the value of bushkit

??

Compared with the issue from where we come from, it's a small one.

wlbragg commented:
@dany93 

>On the runway, with Amphibious.
(I have to change 160 hp to) 180 hp ==> not normal

>Still this 160 hp issue with Floats or Amphibious at launching. Is it difficult to fix? Is it connected to
>>Our save aircraft state is still influx and needs a major overhaul to use the "Startup state" system effectively.

How did you launch, using the variant selection of Amphibian/180? Because mine is working correctly?

I'm not sure if you would be getting false data if you didn't clear the cache after getting the last fix and then testing? 

The automatic changeover from 160 to 180 has always resided in the GUI options. You select float or amphibian and if 160 is selected it sets it to 180. As far as I know that is the only place you can do that and still retain the ability to select the 160 if you want.

You can't ever automatically look to see if float or amph is selected and if so change to 180, because then you have no choice to use 160 and that should be an option.

This next statement fies in face of what currently actually happens as we don't purposely cache the current engine selection...
So how do you remember the setting of 180 across re-inits (reloading), either going to mooring or reloading the next session of the simulator, you have to cache it, we don't but the automatic caching might.
If you explicitly cache it, as our system is currently written, the saved data would override the variant selection choices. That is where part of the reworking of these systems, variant, state, saved-data, comes into play.
The QT-launcher can be thought of as a temporary cache of default startup parameters as well, as in, if you choose Amph/180 in the QT-launcher and don't change it then it will start that way the next time.

@dany93 can you be more specific what actions you didn't think led to the correct configuration?

wlbragg commented:
@wkitty42 

>are you saying that it kept firing the light_manager.init() function? what does that function do?

On every fdm-init it would run that block of nasal. To me the code is unremarkable to any other nasal code so I don't know why it tripped it up and it never used to, it suddenly changed around the time of 2019.x
I can't see why it would cause it to crash

wlbragg commented:
On the decision to remove the offending property of ``` planing-lift-lbs ``` from the JSB function summing
~~~
hydro/buoyancy-lbs
hydro/planing-lift-lbs
~~~
The ``` hydro/planing-lift-lbs ```function actually includes a product of ``` hydro/buoyancy-lbs ``` and other properties. Beings the ``` hydro/active-norm ``` function is a simple on/off value, it is either 0 or greater that 0, in effect. It never uses the value, only the fact that it is greater than 0. Buoyancy is all you need, you either have influence of being on water or you don't. The summing of those two properties make no logical sense to me. Please correct me if I am wrong. In fact because the ``` hydro/planing-lift-lbs ```function actually includes a product of ``` hydro/buoyancy-lbs ``` and you try to sum that again with ``` hydro/buoyancy-lbs ```, I can see a kind of feedback potential, yes? Bottom line is it doesn't add anything to the on/off manner of the function.
Because of this, I think it is not only safe to remove it permanently (VS commenting it out temporarily).

dany93 commented:
@wlbragg 
> How did you launch, using the variant selection of Amphibian/180? Because mine is working correctly?
> can you be more specific what actions you didn't think led to the incorrect configuration?

I usually launch with a .fgfsrc file
` --aircraft=c172p`
However, with Qt and Default (==> same command line` --aircraft=c172p`) I get the same issue.

I think I've found the way to obtain the issue:
**With Amphibious, press "Go Now", then Exit.**
**At the next launching, my aircraft spawns with Amphibious and 160 hp.**
Despite ` --aircraft=c172p`
Only at the _first next_ launching.
The one after this next, without "Go Now", it works correctly (spawns with Default).

As I most of the time tested the "Go Now" during these last days, I almost most of the time got this issue.

Out of that, the result is in accordance with the setting (--aircraft=c172p or --aircraft=c172p-amphibious). I've not tested this many times, but it seems.

onox commented on Nasal/electrical.nas:
```diff
@@ -495,18 +495,26 @@ var avionics_bus_1 = func() {
 
+############################ Utility function
+
+var flapsDown = controls.flapsDown;
```
Where does this code come from?

onox replied:
It seems we used to have an override, but removed it in #1236

onox replied:
(@wlbragg so this change should be removed otherwise we undo the fix of #1236)

wlbragg replied:
@Zaretto indicated this was the proper way to do an overload of a base function in controls, see: https://github.com/c172p-team/c172p/issues/1265#issuecomment-484159150 Although where is the "flapsUp". @gilbertohasnofb  did that commit so I need to let him sort if out if possible.

Zaretto replied:
> @Zaretto indicated this was the proper way to do an overload of a base function in controls, 

That's correct; however it's not necessarily the best way to do what that code is doing and a listener on `controls/flight/flaps` is probably a better way to do it and will handle up and down.

Overriding the methods in $FGData/Nasal/controls.nas is really more suited to prevent a control from being moved due to aircraft specific interlocks. For example on the B737 you can't physically move the gear lever when on the ground (unless you pull a little lever underneath the handle).

onox replied:
@wlbragg But @Zaretto used fgdata. If you check [`master`](https://github.com/c172p-team/c172p/blob/master/Nasal/electrical.nas) you see there's no longer any override, so why introduce it again?

onox commented:
> The only outstanding code is the immat.nas nested listeners that Richard @Zaretto suggested removing. I need verification as to what this code is for and what we need to do to it.

@wlbragg IIRC immat is to display the user's callsign on the fuselage and in the cockpit.

wlbragg commented:
> > The only outstanding code is the immat.nas nested listeners that Richard @Zaretto suggested removing. I need verification as to what this code is for and what we need to do to it.

> @wlbragg IIRC immat is to display the user's callsign on the fuselage and in the cockpit.

I am OK with leaving it as is for now. It isn't causing any problems and as far as I know used in other aircraft. I think it is code that goes way back.

wlbragg commented:
>I think I've found the way to reproduce the issue:
With Amphibious (180 hp), press "Go Now", then Exit.
At the next launching, my aircraft spawns with Amphibious and 160 hp.
Amphibious, despite --aircraft=c172p.
Only at the first next launching.

@dany93 The first time I tried this it worked like you would want it to, it started up with the default tires (all tests with QT launcher). From then on it worked as you said and would only start with default on the load after starting on the runway once. Beings this is not related to this PR or issue of nasal errors and segfaults, I want to get this PR merged. I do think your issue is related to the "state" work that needs to be done.

@gilbertohasnofb if no one has any objections this is ready to be merged. Once it is merged I will contact Stuart and get it merged in to fgdata, or we can cut a release and then ask for it to be pushed to fgdata.

gilbertohasnofb commented:
@wlbragg thank you!

wlbragg commented:
@legoboyvdlp  @gilbertohasnofb Do either of you want to take lead on cutting an actual release?
I can go through the list of new features and work done and get it up dated maybe as early as later tonight. :smile:

gilbertohasnofb commented:
Yeah, I can do that. Since I have been away for quite some time, could you make sure that the list of new features and bug fixes in https://github.com/c172p-team/c172p/issues/1196 is up to date? I think @legoboyvdlp and @dany93 also need to take a look at that.

Also, I see that our last release was version 2018.3 and the current opened issue for the list of new features talks for 2018.4. Can anyone confirm if this is correct? I am suspecting we did not release a 2018.4 and this will actually be a 2019.1 or 2019.2 release. @onox @wlbragg do either of you know what's going on with these version numbers?

legoboyvdlp commented:
That's correct - we want to cut a 2019.1 release as 2018.4 was skipped (hopefully it is in time for a FGDATA release?). I think there are no bugs... it just needs someone to trigger the release. Maybe @zakalawe might know if it's in time to be added to FGDATA for the release?

gilbertohasnofb commented:
Ok, so 2019.1 it is. Please gather the changes for the change log and give me green light once everything has been accounted for so that I can prepare the release.

wlbragg commented:
@gilbertohasnofb  I'll go through the list tonight and ping you when I am done.

gilbertohasnofb commented:
:+1:
